File: library/src/main/java/org/xmtp/android/library/libxmtp/Message.kt
package org.xmtp.android.library.libxmtp

import org.xmtp.android.library.XMTPException
import org.xmtp.android.library.codecs.ContentTypeGroupUpdated
import org.xmtp.android.library.codecs.EncodedContent
import org.xmtp.android.library.codecs.decoded
import org.xmtp.android.library.messages.Topic
import org.xmtp.android.library.toHex
import org.xmtp.proto.message.contents.Content
import uniffi.xmtpv3.FfiConversationMessageKind
import uniffi.xmtpv3.FfiDeliveryStatus
import uniffi.xmtpv3.FfiMessage
import uniffi.xmtpv3.FfiMessageWithReactions
import java.util.Date

class Message private constructor(
    private val libXMTPMessage: FfiMessage,
    val encodedContent: Content.EncodedContent,
    private val decodedContent: Any?,
    val childMessages: List<Message>? = null
) {
    enum class MessageDeliveryStatus {
        ALL, PUBLISHED, UNPUBLISHED, FAILED
    }

    enum class SortDirection {
        ASCENDING,
        DESCENDING;
    }

    val id: String
        get() = libXMTPMessage.id.toHex()

    val convoId: String
        get() = libXMTPMessage.convoId.toHex()

    val senderInboxId: String
        get() = libXMTPMessage.senderInboxId

    val sentAt: Date
        get() = Date(libXMTPMessage.sentAtNs / 1_000_000)

    val sentAtNs: Long
        get() = libXMTPMessage.sentAtNs

    val deliveryStatus: MessageDeliveryStatus
        get() = when (libXMTPMessage.deliveryStatus) {
            FfiDeliveryStatus.UNPUBLISHED -> MessageDeliveryStatus.UNPUBLISHED
            FfiDeliveryStatus.PUBLISHED -> MessageDeliveryStatus.PUBLISHED
            FfiDeliveryStatus.FAILED -> MessageDeliveryStatus.FAILED
        }

    val topic: String
        get() = Topic.groupMessage(convoId).description

    @Suppress("UNCHECKED_CAST")
    fun <T> content(): T? = decodedContent as? T

    val fallbackContent: String
        get() = encodedContent.fallback

    val body: String
        get() {
            return content() as? String ?: fallbackContent
        }

    companion object {
        fun create(libXMTPMessage: FfiMessage): Message? {
            return try {
                val encodedContent = EncodedContent.parseFrom(libXMTPMessage.content)
                if (encodedContent.type == ContentTypeGroupUpdated && libXMTPMessage.kind != FfiConversationMessageKind.MEMBERSHIP_CHANGE) {
                    throw XMTPException("Error decoding group membership change")
                }
                // Decode the content once during creation
                val decodedContent = encodedContent.decoded<Any>()
                Message(libXMTPMessage, encodedContent, decodedContent)
            } catch (e: Exception) {
                null // Return null if decoding fails
            }
        }

        fun create(libXMTPMessageWithReactions: FfiMessageWithReactions): Message? {
            return try {
                val encodedContent = EncodedContent.parseFrom(libXMTPMessageWithReactions.message.content)
                if (encodedContent.type == ContentTypeGroupUpdated && libXMTPMessageWithReactions.message.kind != FfiConversationMessageKind.MEMBERSHIP_CHANGE) {
                    throw XMTPException("Error decoding group membership change")
                }
                // Decode the content once during creation
                val decodedContent = encodedContent.decoded<Any>()

                // Convert reactions to Message objects
                val reactionMessages = libXMTPMessageWithReactions.reactions.mapNotNull { create(it) }

                Message(
                    libXMTPMessageWithReactions.message,
                    encodedContent,
                    decodedContent,
                    reactionMessages
                )
            } catch (e: Exception) {
                null // Return null if decoding fails
            }
        }
    }
}

File: Sources/XMTPiOS/Libxmtp/Message.swift
import Foundation
import LibXMTP

enum MessageError: Error {
	case decodeError(String)
}

public enum MessageDeliveryStatus: String, Sendable {
	case all
	case published
	case unpublished
	case failed
}

public enum SortDirection {
	case ascending
	case descending
}

public struct Message: Identifiable {
	let ffiMessage: FfiMessage
	private let decodedContent: Any?
    public let childMessages: [Message]?

	public var id: String {
		ffiMessage.id.toHex
	}

	public var convoId: String {
		ffiMessage.convoId.toHex
	}

	public var senderInboxId: String {
		ffiMessage.senderInboxId
	}

	public var sentAt: Date {
		Date(
			timeIntervalSince1970: TimeInterval(ffiMessage.sentAtNs)
				/ 1_000_000_000)
	}

	public var sentAtNs: Int64 {
		ffiMessage.sentAtNs
	}

	public var deliveryStatus: MessageDeliveryStatus {
		switch ffiMessage.deliveryStatus {
		case .unpublished:
			return .unpublished
		case .published:
			return .published
		case .failed:
			return .failed
		}
	}

	public var topic: String {
		Topic.groupMessage(convoId).description
	}

	public func content<T>() throws -> T {
		guard let result = decodedContent as? T else {
			throw MessageError.decodeError(
				"Decoded content could not be cast to the expected type \(T.self)."
			)
		}
		return result
	}

	public var fallbackContent: String {
		get throws {
			try encodedContent.fallback
		}
	}

	public var body: String {
		get throws {
			do {
				return try content() as String
			} catch {
				return try fallbackContent
			}
		}
	}

	public var encodedContent: EncodedContent {
		get throws {
			try EncodedContent(serializedBytes: ffiMessage.content)
		}
	}

	public static func create(ffiMessage: FfiMessage)
		-> Message?
	{
		do {
			let encodedContent = try EncodedContent(
				serializedBytes: ffiMessage.content)
			if encodedContent.type == ContentTypeGroupUpdated
				&& ffiMessage.kind != .membershipChange
			{
				throw MessageError.decodeError(
					"Error decoding group membership change")
			}
			// Decode the content once during creation
			let decodedContent: Any = try encodedContent.decoded()
			return Message(
                ffiMessage: ffiMessage, decodedContent: decodedContent, childMessages: nil)
		} catch {
			print("Error creating Message: \(error)")
			return nil
		}
	}
    
    public static func create(ffiMessage: FfiMessageWithReactions)
        -> Message?
    {
        do {
            let encodedContent = try EncodedContent(
                serializedBytes: ffiMessage.message.content)
            if encodedContent.type == ContentTypeGroupUpdated
                && ffiMessage.message.kind != .membershipChange
            {
                throw MessageError.decodeError(
                    "Error decoding group membership change")
            }
            // Decode the content once during creation
            let decodedContent: Any = try encodedContent.decoded()

			let childMessages = try ffiMessage.reactions.map { reaction in
                let encodedContent = try EncodedContent(
                    serializedBytes: reaction.content)
                // Decode the content once during creation
                let decodedContent: Any = try encodedContent.decoded()
                return Message(ffiMessage: reaction, decodedContent: decodedContent, childMessages: nil)
			}

            return Message(
                ffiMessage: ffiMessage.message, decodedContent: decodedContent, childMessages: childMessages)
        } catch {
            print("Error creating Message: \(error)")
            return nil
        }
    }
}
